### ISSUE TITLE
Add restriction of valid values based on selection of other added values

### ISSUE THREAD

**Issue by milen-sage** [Sage-Bionetworks/schematic#179]
Currently we have data model dependencies requiring attributes depending on selected values for other attributes. However, another use case is restricting the values that can be selected for one attribute, based on a selection of values for another attribute.

We can handle these additional DAG dependencies in different ways on the backend, but need to specify an implementation. However, this is yet another consideration for the front end design... e.g. how do we handle that w/in google sheets - i.e. setting valid values dynamically?

**Comment by Aryllen** [comment]
We have a use case for AD. For example, if we have `assay == rnaSeq`, then we want to enforce that `dataType == geneExpression`.

**Comment by allaway** [comment]
+1 to this feature

**Comment by ychae** [comment]
@milen-sage sure! I'll ping a couple of the users and put the results in the comments here.

**Comment by allaway** [comment]
How important is this feature?
• Low-Medium - can do work without it; but it’s important (e.g. to save time or for convenience) - mainly to save time for our users :)

When will use cases depending on this become relevant?
• Short-term - 2-4 weeks - relevant to manifests we are using right now. 

Let me know if you need any more info!

**Comment by ychae** [comment]
From @Aryllen:

**Is your feature request related to a problem? Please describe.**
Data contributors may not understand the relationships between our terms. Contributors have previously annotated raw rnaSeq data with `resourceType = analysis` instead of `experimentalData` . Trying to find all of the incorrect pairings in tens of thousands of files by hand is not scalable.
**Describe the solution you’d like**
Ability to generate schemas that check for enforced pairings between terms. In the terms of the spreadsheet, I am not sure what that would look like. However, there would need to be a way to say that if the value of 'hocus' is 'pocus', then the only allowable terms for 'witches' are 'Winifred', 'Mary', and 'Sarah'. A value of 'Glinda' for 'witches' would be invalid and able to be flagged by the schema.
**How important is this feature?** 
• Medium - can do work without it; but it’s important (e.g. to save time or for convenience)
**When will use cases depending on this become relevant? Select from the options below:**
• Short-term - 2-4 weeks
For AD, I would say short-term since we already have use-cases -- we need to audit the annotations on the AD Knowledge Portal data. However, it is potentially looking like we will need to hand-craft our schemas instead of using this tool anyway (at least for the near future since there isn't the ability to require values based on boolean values). In that sense, there is less urgency unless there are people who need this feature and are already planning on using this tool.
**Additional context**
Cannot think of anything else related to this feature at the moment.

**Comment by milen-sage** [comment]
@Aryllen thanks! This is useful information to know independent of whether APM-AD will end up using schematic.

**Comment by BrunoGrandePhD** [comment]
I'm thinking of how we could implement this feature, at least at the level of the CSV data model specification. 

I'll use the same example as in #214. Briefly, we have a model for clinical data where the unconditionally required columns are `PatientId`, `Diagnosis`, and `Sex`; an entirely optional column is `FavoriteColor`; and a `Cancer` diagnosis requires `CancerType` and `DiagnosisDate`, and in turn, a `Skin` cancer type conditionally requires `CancerClassification` since so many skin cancers are benign and should be considered separately. I've included the corresponding CSV data model below. Note that `DependsOn` was formerly called `Requires` (with an `s`). 

```
Attribute            ,Valid Values         ,DependsOn                               ,Required
---------------------|---------------------|----------------------------------------|--------
ClinicalData         ,                     ,"PatientId,Diagnosis,Sex,FavoriteColor" ,FALSE
Diagnosis            ,"Cancer,Healthy"     ,                                        ,TRUE
PatientId            ,                     ,                                        ,TRUE
Sex                  ,"Female,Male,Other"  ,                                        ,TRUE
FavoriteColor        ,"Red,Green,Blue"     ,                                        ,FALSE
Cancer               ,                     ,"CancerType,DiagnosisDate"              ,FALSE
CancerType           ,"Skin,Lung,Colon"    ,                                        ,TRUE
DiagnosisDate        ,                     ,                                        ,TRUE
Skin                 ,                     ,CancerClassification                    ,FALSE
CancerClassification ,"Benign,Malignant"   ,                                        ,TRUE
```

In my mental model of these CSV specifications, I group these rows into one of three categories (but there are others not shown here): 

1. A "data type" (_e.g._ `ClinicalData`)
2. A "data type attribute" (_e.g._ `PatientId`, `Diagnosis`, `CancerType`), not to be confused with the way we use "attribute" to refer to each row in this CSV specification
3. A "valid value" (_e.g._ `Cancer`, `Skin`)

Like Milen described, we currently allow a "valid value" to require a "data type attribute". In our example above, this was achieved when the `Cancer` "valid value" was assigned `CancerType,DiagnosisDate` under `DependsOn`. 

In theory, I don't see why a "valid value" couldn't also require a "valid value" (or one of multiple "valid values") for another "data type attribute". For instance, let's say that we want to require that `CancerClassification` be set to `Malignant` if `CancerType` is set to `Lung` or `Colon`. The question then becomes, how do we capture this relationship in the CSV format? 

One option is listing the required "valid value(s)" (_e.g._ `Malignant`) under the `Valid Values` for the "valid value" in question (_e.g._ `Lung` and `Colon`). (What a sentence... I'll elaborate with an example CSV below.) This would imply that the values listed under `Valid Values` are the only compatible values under their respective "data type attributes" (_e.g._ `CancerClassification`). By using `Valid Values` in this way, it would allow a given "valid value" to simultaneously and unambiguously require "data type attributes" (under `DependsOn`) and "valid values" for other "data type attributes" (under `Valid Values`). 

One assumption I'm making here is that a "valid value" will never share the same name with a "data type attribute". For example, it wouldn't be possible to make `Cancer` a "data type attribute" by including it under `DependsOn` for the `ClinicalData` "data type", because the use of the `Valid Values` column for `Cancer` represents something different than for your typical "data type attribute". 

To illustrate my proposed approach, I've updated the CSV data model specification below to capture this requirement: `CancerClassification` must be set to `Malignant` if `CancerType` is set to `Lung` or `Colon`. Specifically, I added the `Lung` and `Colon` rows. 

```
Attribute            ,Valid Values        ,DependsOn                               ,Required
---------------------|--------------------|----------------------------------------|--------
ClinicalData         ,                    ,"PatientId,Diagnosis,Sex,FavoriteColor" ,FALSE
PatientId            ,                    ,                                        ,TRUE
Diagnosis            ,"Cancer,Healthy"    ,                                        ,TRUE
Cancer               ,                    ,"CancerType,DiagnosisDate"              ,FALSE
CancerType           ,"Skin,Lung,Colon"   ,                                        ,TRUE
Skin                 ,                    ,CancerClassification                    ,FALSE
Lung                 ,Malignant           ,                                        ,FALSE
Colon                ,Malignant           ,                                        ,FALSE
CancerClassification ,"Benign,Malignant"  ,                                        ,TRUE
DiagnosisDate        ,                    ,                                        ,TRUE
Sex                  ,"Female,Male,Other" ,                                        ,TRUE
FavoriteColor        ,"Red,Green,Blue"    ,                                        ,FALSE
```

Is this dual purpose of the `Valid Values` column intuitive? I'm just trying to minimize the number of columns that are added to the CSV format so that it doesn't become too unwieldy.

**Comment by milen-sage** [comment]
> One option is listing the required "valid value(s)" (e.g. Malignant) under the Valid Values for the "valid value" in question (e.g. Lung and Colon).
Shouldn't the universe implode when a sentence that has never been uttered and should have been unutterable is uttered? :)

**Comment by milen-sage** [comment]
@BrunoGrandePhD I see the proposed logic, to make it more precise wouldn't we want

```
Attribute            ,Valid Values        ,DependsOn                               ,Required
---------------------|--------------------|----------------------------------------|--------
ClinicalData         ,                    ,"PatientId,Diagnosis,Sex,FavoriteColor" ,FALSE
PatientId            ,                    ,                                        ,TRUE
Diagnosis            ,"Cancer,Healthy"    ,                                        ,TRUE
Cancer               ,                    ,"CancerType,DiagnosisDate"              ,FALSE
CancerType           ,"Skin,Lung,Colon"   ,                                        ,TRUE
Skin                 ,                    ,CancerClassification                    ,FALSE
Lung                 ,Malignant           ,**CancerClassification**                                        ,FALSE
Colon                ,Malignant           ,**CancerClassification**                                        ,FALSE
CancerClassification ,"Benign,Malignant"  ,                                        ,TRUE
DiagnosisDate        ,                    ,                                        ,TRUE
Sex                  ,"Female,Male,Other" ,                                        ,TRUE
FavoriteColor        ,"Red,Green,Blue"    ,                                        ,FALSE
```

That way we'd know that the restriction of valid values based on Lung or Colon applies to CancerClassification? This opens up the question what if we have multiple attributes in 'DependsOn' for 'Lung' (or 'Colon'). In that case, we should check if the valid values for Lung/Colon are a subset of the valid values in each attribute in the DependsOn column - if so, we apply the Lung/Colon valid values subset restriction?

**Comment by BrunoGrandePhD** [comment]
From what I've gathered so far, it seems like we aim to minimize the number of required "data type attributes" when designing data models. Given this preference, I thought it would be helpful to make it optional to list the associated "data type attribute" in the `DependsOn` column since in some cases (like in my example above), you might not care as much to know about `CancerClassification` if it's lung or colon cancer (unlike if it's skin cancer). 

And like you said, including the associated "data type attribute" under `DependsOn` doesn't really solve the issue of knowing where the "valid value" comes from. When I wrote my comment, I was thinking that the associated "data type attribute" could be determined by looking it up in the graph, but (1) I'm not sure whether "valid values" are added as nodes in the graph, and (2) I don't think we can assume that a "valid value" is uniquely tied to a "data type attribute". @milen-sage: Could you enlighten me here? 

If "valid values" cannot be uniquely tied to a "data type attribute", then we indeed need a mechanism to specify the "data type attribute" being referred to unambiguously. Using the approach described by Milen does dramatically reduce the "search space", but we still cannot guarantee lack of ambiguity if there are two or more "data type attributes" listed in `DependsOn`. Are we comfortable with that tiny probability? 

One possible solution is to ensure that the list in `DependsOn` is as long as the list in `Valid Values`, and the corresponding items from each list are related. Though, this wouldn't allow users to choose whether to make the associated "data type attribute" required or not (as I described in the first paragraph here). 

Alternatively, we could format `Valid Values` as something like `CancerClassification.Malignant`, which would be unambiguous, but that introduces a new syntax that's not found anywhere else in the CSV data model.

**Comment by milen-sage** [comment]
> From what I've gathered so far, it seems like we aim to minimize the number of required "data type attributes" when designing data models. Given this preference, I thought it would be helpful to make it optional to list the associated "data type attribute" in the `DependsOn` column since in some cases (like in my example above), you might not care as much to know about `CancerClassification` if it's lung or colon cancer (unlike if it's skin cancer).
> 
> And like you said, including the associated "data type attribute" under `DependsOn` doesn't really solve the issue of knowing where the "valid value" comes from. When I wrote my comment, I was thinking that the associated "data type attribute" could be determined by looking it up in the graph, but (1) I'm not sure whether "valid values" are added as nodes in the graph, and (2) I don't think we can assume that a "valid value" is uniquely tied to a "data type attribute". @milen-sage: Could you enlighten me here?
> 

1) Yes, all valid values are added as nodes in the graph
2) A valid value is not uniquely tied to a data type attribute (i.e. it could be reused as a valid value in other data type attributes)

> If "valid values" cannot be uniquely tied to a "data type attribute", then we indeed need a mechanism to specify the "data type attribute" being referred to unambiguously. Using the approach described by Milen does dramatically reduce the "search space", but we still cannot guarantee lack of ambiguity if there are two or more "data type attributes" listed in `DependsOn`. Are we comfortable with that tiny probability?
> 

In the schemas we have now, we haven't bumped yet in a case where we this ambiguity would arise. So, I'd be ok with giving this a try. If there ambiguity we can throw an error that the ambiguity needs to be resolved (e.g. by renaming valid values). In the near future we can be fool-proof by exposing the jsonld schema context (i.e. in "rdfs:label" rdfs is a context; in "bts:DataTypeAttribute" bts is a context, etc.) for each attribute as a column in the csv model definition. So if Lung, the valid values of Lung, and CancerClassification have the same context, but the context(s) of other attributes specified in Lung's 'DependsOn' values is(are) different, that could be used as disambiguation. This is similar to the proposal below for having syntax like CancerClassification.Malignant, except the context syntax is already supported in json-ld.

> One possible solution is to ensure that the list in `DependsOn` is as long as the list in `Valid Values`, and the corresponding items from each list are related. Though, this wouldn't allow users to choose whether to make the associated "data type attribute" required or not (as I described in the first paragraph here).
> 
> Alternatively, we could format `Valid Values` as something like `CancerClassification.Malignant`, which would be unambiguous, but that introduces a new syntax that's not found anywhere else in the CSV data model.

**Comment by BrunoGrandePhD** [comment]
I agree that a warning (or even an error) in the unlikely case of an ambiguity would be essential. 

Alright, I think we have a way forward with this issue.

**Comment by ychae** [comment]
Moving this to Backlog for this phase as its associated issue is also in Backlog. @BrunoGrandePhD please move if I'm incorrect!

**Comment by BrunoGrandePhD** [comment]
This is a nice-to-have feature for imCORE. 

@milen-sage: Is this the issue that's intended to track the context feature? You mention the use of contexts in your last comment, and I couldn't find an issue when searching for "context". Should that be tracked separately? Or will tracking it here with a specific use case suffice?

**Comment by milen-sage** [comment]
@BrunoGrandePhD I haven't created the issue for context, yet. I'll start it later and link to it here.

**Comment by danlu1** [comment]
+1 to this feature for SysBio DCC projects. One example is that our data has different assays associated with different dataType. Currently, the assay drop-down menu is very long (104 assays for PEC). If we can shorten the drop-down menu and only show the assay corresponding to the specific dataType, that would empower the user in filling metadata.

**Comment by milen-sage** [comment]
@ychae let's triage this one. It may be an epic since it requires a couple of dependency issues we need to outline.

**Comment by AmyHeiser** [comment]
Adding some additional comments shared on Slack for reference: 
From Robert on 9/6/23: 
For example, it doesn’t seem like we can dynamically select valid values based on the selection in another attribute (e.g. only list genomics assays if a user selects dataType = ‘genomics’), but I was curious if this could be handled at least at validation somehow (e.g. if someone selects dataType = ‘genomics’ and then a non-genomics assay, validation fails)

From Anh Nguyet: 

1. Related to the above and should be easier to implement, how about a feature for "submit as"? So for example under platform we have a very long list of platforms. We are OK with specifically defining sequencingPlatform , imagingPlatform , etc. to help constrain platform to ones that make sense for the assay type, but if we do this the metadata gets submitted with sequencingPlatform key and is not useful for the portal fileview filter on platform.
2. Manifests should only pull in properties defined in that manifest when there are existing annotations. For example, we add some extra metadata in an automated manner to files (initiative, fundingAgency, etc.). These get pulled in when someone is editing file metadata using a specific template, but they're not relevant, and it makes the template look very busy.
3. Make it easer to apply optional additional metadata to files. For example, let's say imaging files undergo annotation using the main imaging data template. However, some people may want to add extra metadata to the files -- like which researcher did the acquisition, associated milestone, additional links and comments  -- basically things that would normally on a different sheet and should be added as a separate step/"supplement" template. Using Component disallows this. Also depends on #2 being resolved.

**Comment by AmyHeiser** [comment]
Based on team discussion - it would be easier to do on a manifest by manifest basis vs. on a row by row basis within one manifest. If based on the Project/folder options you could designate which manifest would be generated.   @allaway @anngvu

**Comment by anngvu** [comment]
@AmyHeiser By-template is what makes sense and what I had in mind.

The way that the schematic data model works, you can't define a different set of values for a property of the same name. When HTAN wants to constrain values, for example, they do something like use different properties such as `AssayType` vs `ImagingAssayType`, where the imaging-specific one appears in the imaging template for imaging data. Again, we can't do this across our templates because then it wouldn't work for the portal. So we could either use the "alias" solution or other way to specify what the subset should be depending on the **template context**.

The design goal is to have easier and less-likely-to-be-incorrect selections for the user. I think there are others who might still want the row-by-row version, though, so even though I'd be OK with template-level stuff, some others might not be?

Also, is there a feature request somewhere for defining default values? It would also save a lot of time for some things - e.g. file formats.
